### PR TITLE
Set aimer voltage to small negative when near the hard stop

### PR DESCRIPTION
**Pull request by minhnguyenbhs**

### CODE REVIEW

**Review by jkleiber — APPROVED** [review]
Looks good, if this was tested on the robot please merge

**Comment by linglejack06** [comment]
@jkleiber @minhnguyenbhs id like to test this in shop today before you merge it. I say this because I noticed yesterday that when moving the robot the inertia causes the Aimer to go up and down. Therefore the aimer doesn't stay at its bottom position and can't intake a note well. We retuned aimer PId to make it stay at set point a little better. I'm not sure how well the aimer will stay there at zero voltage.

**Comment by jkleiber** [comment]
This might be something we do based on mode (PID hold to 0 during intake, let sit during idle)

In any case I dismissed my review until this can be tested

**Comment by minhnguyenbhs** [comment]
Are we confirmed to have shop today from 10-1?Minh NguyenOn Oct 17, 2024, at 8:36 AM, Justin Kleiber ***@***.***> wrote:﻿
This might be something we do based on mode (PID hold to 0 during intake, let sit during idle)
In any case I dismissed my review until this can be tested

**Comment by PChild** [comment]
> Are we confirmed to have shop today from 10-1? @linglejack06

Yes, I'm opening from 10-1.

**Comment by linglejack06** [comment]
> This might be something we do based on mode (PID hold to 0 during intake, let sit during idle)
> 
> In any case I dismissed my review until this can be tested

Yeah, that's how it's set up now. For whatever reason about 1 in 5 times the intake would shoot note before aimer was all the way down ( hasn't been tested with new PId though) this would cause the note to go above aimer and get stuck in robot. Minh or I will test it out and let you know

**Comment by aidnem** [comment]
@linglejack06 was this the case after the encoder got fixed as well? It could've been the arm thinking it was down and allowing the note to pass through when it was actually up.

**Comment by linglejack06** [comment]
@jkleiber we ran out of time to test this, but should be quick to test before comp on Saturday

**Comment by aidnem** [comment]
@minhnguyenbhs could we update this to apply a small negative voltage and lower the current limit when arm is near the bottom and we are intaking? This way, it will gently drive the arm into the hardstop and keep it snapped down even during acceleration.

**Comment by minhnguyenbhs** [comment]
ok! will get to that tonight

File: src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java:241]: As written, this would slowly bring the aimer back to the intake location. We only want to apply this small output if the goal AND the aimer position are both near the hard stop

However this might mess with long shots as written. So maybe rather than using the goalAngleRot we could set a flag for if we want to drive into the hard stops and turn that on and off from shooter subsystem

Comment (minhnguyenbhs) [reply src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java:241]: Should I change this to be activated for state == intake when goalAngleRotation is close to zero and actual angle is close to zero, then? Or would it be better to create a new flag of its own?

Comment (jkleiber) [reply src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java:241]: I'd prefer a new flag just to decouple it from states.

For example we may want to set this flag true in idle as well

**Review by jkleiber — APPROVED** [review]

**Pull request merged**
